index.js

```diff
@@ -283,6 +294,126 @@ class Conf {
 		this.store = store;
 	}
 
+	_serializeExtraTypes(value) {
+		// Base cases
+		for (const type of this.extraTypes) {
+			if (type.isInstance(value)) {
+				return {
+					[TYPE_KEY]: type.name,
+					value: type.convertFrom(value)
+				};
+			}
+		}
+
+		// Value is a base type
+		if (typeof value !== 'object') {
+			return value;
+		}
+
+		// Sub-object handling of extra types
+		const stack = [];
+		const objectSeenList = [];
+
+		stack.push({current: value, parent: null, key: null});
+		let circularReferenceCounter = 1;
+		while (stack.length > 0) {
+			const {current, parent, key} = stack.pop();
+			if (objectSeenList.indexOf(current) === -1) {
+				objectSeenList.push(current);
+			} else {
+				// Current object is reference to an object already parsed
+				// current is the circular reference
+				// seenList.indexOf(current) is the object it's pointing to
+				const referenceTarget = objectSeenList[objectSeenList.indexOf(current)];
+				parent[key] = {
+					[TYPE_KEY]: 'circularObject',
```

Comment:
I don't really see why you need the `circularObject` thing. Circular objects can simply be detected by comparing references.  I think this logic could be simplified a lot.

See how it's done here: https://github.com/sindresorhus/serialize-error/blob/master/index.js and here: https://github.com/sindresorhus/map-obj/blob/master/index.js

Reply:
The `circularObject` type signals to the deserialize function that it was a circular object, so it can lookup the `referenceNumber` set in the value of the object, and set it to the object it was referencing to before serialization.

I might have misunderstood this problem. I made it so that when deserializing the circulation is reconstructed, so you're going to end up with the same circular object after getting it from the store.
So should I just completely remove circulation like the `destroyCircular` function does in the first example you've provided, or is there an easier way to reconstruct the circulation?